feat: Enhance private link expiration with usage and date limits

retrogtx opened the pull request:
## What does this PR do?

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

- Fixes #19906 (GitHub issue number)
- Fixes CAL-5283 (Linear issue number - should be visible at the bottom of the GitHub issue description)

https://www.loom.com/share/70bf41049d7a4e36b8f5a894e8647739?sid=3a37fce8-5c18-441b-8640-bbc6e202a661

## Mandatory Tasks (DO NOT REMOVE)

- [ ] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox. N/A.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

retrogtx committed:
feat: Enhance private link expiration with usage and date limits

retrogtx committed:
fix: prisma enum error

asadath1395 commented:
@retrogtx You beat me on this :D. I was almost done with the implementation, i had implemented it differently as in instead of dropdown, just provided the max usage and expiry date field as input and then based on the values add the checks for expiry and also show the link expiry error message below the link. Anyways good job!

retrogtx commented:
hey @asadath1395 

thanks for trying, there are many other issues lying around along with bounties that might interest you

thank you again 🫡

retrogtx committed:
add tests

retrogtx committed:
Merge branch 'main' into priv-links

alishaz-polymath reviewed (changes requested):
Clean work! 🙏 
There's a NIT though, I'd rather have the expired links showing still and shown as expired instead of turning off the private links automatically upon expiration. Also, it would be better to have a visual representation of how many usages are left for the private link (or when expiration is due) for example. 
Otherwise, there is a great overhead in support waiting to crash!

retrogtx committed:
feat: Enhance hashed link functionality with usage tracking and real-time updates

- Added usageCount to PrivateLinkWithOptions type for tracking link usage.
- Updated getServerSideProps to select hashed link data including usageCount.
- Refactored MultiplePrivateLinksController to fetch and display real-time usage data for links.
- Implemented cache invalidation on component mount to ensure fresh data retrieval.
- Improved link description logic to reflect usage limits and expiration status accurately.

retrogtx committed:
Merge branch 'main' into priv-links

retrogtx committed:
attempt to fix e2e pls work

retrogtx commented:
🫡 @alishaz-polymath everything's done

retrogtx committed:
- validation to ensure maxUsageCount is greater than current usage count for private links.
- Added error messages for usage limit reached and invalid max usage count.
- Updated MultiplePrivateLinksController to handle validation and display appropriate error notifications.

retrogtx committed:
make e2e test pass again

alishaz-polymath reviewed (dismissed):
A few things failing:
1. When I try to create multiple private links, say I have created 1 that expires on a certain date -> As soon as I try to "Add new link", I get error: `Error: Rendered more hooks than during the previous render.`
2. The expiration date is not working. My expiration date is set to 15 of March 2025, yet I am able to access the booking page on 18th of March 2025
3. Translation missing for `link_expired_on_date`
<img width="966" alt="image" src="https://github.com/user-attachments/assets/67f03006-0ac1-4f64-bb9c-4c2638fe24bc" />

retrogtx commented:
That weird since one can't choose a previous date in the first place

<img width="1512" alt="image" src="https://github.com/user-attachments/assets/d28ee12c-a243-4e36-960a-dee05cd7b7a5" />

alishaz-polymath commented:
> That weird since one can't choose a previous date in the first place
> 
> <img alt="image" width="1512" src="https://private-user-images.githubusercontent.com/76801862/423919867-d28ee12c-a243-4e36-960a-dee05cd7b7a5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIyOTI5ODUsIm5iZiI6MTc0MjI5MjY4NSwicGF0aCI6Ii83NjgwMTg2Mi80MjM5MTk4NjctZDI4ZWUxMmMtYTI0My00ZTM2LTk2MGEtZGVlMDVjZDdiN2E1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzE4VDEwMTEyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkyMTc5NjQzY2JlNGJkNzhkOTU1Mjc0OWQ1OTI4YzEzMThiZGM5MWVlM2U0MWEzMjViYjYzZWFlNTY0NDRkMDUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.vG1WSTG5eSLD9yr2VXOBm_gsRkbFrSdatsVDLc7-slI">

Yeah I updated the date directly in the DB for testing purposes.

retrogtx committed:
feat: Implement link expiration and usage limit checks for date expiry

- Added logic in getServerSideProps to block access if the link is expired or has exceeded its usage limit.
- Updated common.json to include a message for expired links with a specific date.
- Refactored HashedLinkUsageIndicator and MultiplePrivateLinksController to handle bulk fetching of hashed links and improve data handling.
- Introduced getHashedLinks endpoint in the TRPC router for batch retrieval of hashed link data.

retrogtx commented:
btw it has been implemented, done! @alishaz-polymath

alishaz-polymath commented:
Looks 

> A few things failing:
> 
> 1. When I try to create multiple private links, say I have created 1 that expires on a certain date -> As soon as I try to "Add new link", I get error: `Error: Rendered more hooks than during the previous render.`
> 2. The expiration date is not working. My expiration date is set to 15 of March 2025, yet I am able to access the booking page on 18th of March 2025
> 3. Translation missing for `link_expired_on_date`
> 
> <img alt="image" width="966" src="https://private-user-images.githubusercontent.com/52925846/423917231-67f03006-0ac1-4f64-bb9c-4c2638fe24bc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIzMDI2NTQsIm5iZiI6MTc0MjMwMjM1NCwicGF0aCI6Ii81MjkyNTg0Ni80MjM5MTcyMzEtNjdmMDMwMDYtMGFjMS00ZjY0LWJiOWMtNGMyNjM4ZmUyNGJjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzE4VDEyNTIzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThlMDBjNmUyN2UzNTc0MDlkNzAwNGFmNjRlY2MzOTY3MjI2NTg3YzJkMTEyZmMzNzk2ZDg0ZjBiNTIyNWI2MjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.6PzV3xpbU0VjZKIGU6g09sPBLT4ssgVBxwKWNPz3b3Y">

These issues seem to be addressed 🙏 
Reviewing the code right now, but I am wondering if the flow where the expired private link leads to private link toggle being turned off is the desired flow here.
@CarinaWolli Can you confirm if that's what we want?

CarinaWolli reviewed (commented):
I have some UI suggestions. I think it would look cleaner if we don't show the 'Expires on' and 'Maximum number of uses' inputs and only have the gray text there:

![Screenshot 2025-03-19 at 9 51 17 AM](https://github.com/user-attachments/assets/5940f7e0-9b2f-4c11-82b3-771753c18106)

Instead, open a Dialog after clicking the settings button and handle all settings in there. In the dialog you can either pick 'Usage-based or 'Time-based', by default it's a usage-based linked with Number of uses = 1

Let me know what you think about that

alishaz-polymath commented:
> I have some UI suggestions. I think it would look cleaner if we don't show the 'Expires on' and 'Maximum number of uses' inputs and only have the gray text there:
> 
> ![Screenshot 2025-03-19 at 9 51 17 AM](https://private-user-images.githubusercontent.com/30310907/424384927-5940f7e0-9b2f-4c11-82b3-771753c18106.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIzNzUxNDMsIm5iZiI6MTc0MjM3NDg0MywicGF0aCI6Ii8zMDMxMDkwNy80MjQzODQ5MjctNTk0MGY3ZTAtOWIyZi00YzExLTgyYjMtNzcxNzUzYzE4MTA2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzE5VDA5MDA0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFkYjI5MjJkNTJhZjc3ODcwYjQxMjE2NmVjOTUyNmI1OTUyZjRkZTU2ZWRlNzY0YzgwMDZmZTM2OTA0NzU0ZDUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.RVl7zaCG41e8FYYDnYMVR5pQuP_qOKlAMMRvybz0nHs)
> 
> Instead, open a Dialog after clicking the settings button and handle all settings in there. In the dialog you can either pick 'Usage-based or 'Time-based', by default it's a usage-based linked with Number of uses = 1
> 
> Let me know what you think about that

If I understand you correctly, the dialog where the user selects "time based" or "usage based", they get the exact number/date input there so that we don't see an input afterwards, right? I agree that's a lot cleaner yeah. Love it!
@retrogtx what do you think?

retrogtx commented:
will make the changes 🫡

retrogtx committed:
feat: Add link settings dialog

- Introduced a dialog for configuring link settings, allowing users to set expiration dates and usage limits.
- Refactored MultiplePrivateLinksController to manage link types and settings more effectively.
- Updated common.json to include a new translation for "Link Settings".

alishaz-polymath reviewed (dismissed):
<img width="780" alt="image" src="https://github.com/user-attachments/assets/debfd3fd-933f-489a-91c2-65f7773920fd" />
The alignment is off for the radio button and respective labels (think there's a `mb-2` in there causing it)

Also, I can select 20th march (today) as the date when it should expire, and yet I see the message "This link expired on 20th March". I think there's a UX issue in there somewhere because if I can select the expiration right now I should be able to access the link right now, it shouldn't say "expired" yet. What actually happens is that I can select today from the datepicker, but then it says the link "expired" today, and if I copy paste the link, it leads to 404 meaning it has expired. Thus, it's not great UX.
<img width="960" alt="image" src="https://github.com/user-attachments/assets/9924dddc-a79b-4c99-b750-4ca699ae2856" />


Everything else seems to be good IMO.

CarinaWolli commented:
Can we use the same radio buttons as here? 
<img width="517" alt="Screenshot 2025-03-20 at 12 07 40 PM" src="https://github.com/user-attachments/assets/89e19c32-6420-4602-8e8d-41dec5dc775a" />

Also I would remove 'single use' as an option and only have 'time-based' and 'usage-based'. The default is usage-based with 1 as the input

retrogtx commented:
on it

retrogtx commented:
<img width="1608" alt="image" src="https://github.com/user-attachments/assets/0377fc98-ca07-4f19-aa1e-5f8fd3cf15bf" />
fixed, will push everything in a while!

alishaz-polymath commented:
> <img alt="image" width="1608" src="https://private-user-images.githubusercontent.com/76801862/424967945-0377fc98-ca07-4f19-aa1e-5f8fd3cf15bf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI0NzI0NzcsIm5iZiI6MTc0MjQ3MjE3NywicGF0aCI6Ii83NjgwMTg2Mi80MjQ5Njc5NDUtMDM3N2ZjOTgtY2EwNy00ZjE5LWFhMWUtNWY4ZmQzY2YxNWJmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzIwVDEyMDI1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThjYjUzZDljMjVlOThkNTMzNWM4YzVhYTc4ODI2MmFmYjNhMmY2M2VlOGQ3MDc5NjA0NDkyNDRmZDdjN2Q1ZjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.MDD7MhejD5YPc6Qi5GNX0b9js6FM-sCwD22EQFqScXk"> fixed, will push everything in a while!

![image](https://github.com/user-attachments/assets/4ac22b5c-c28e-43cb-823b-88e273f036c0)
 Can we not have the input inline like so?

retrogtx committed:
fix: the links expire today midnight if today's date is selected

alignment of circular checks along with the locale

retrogtx commented:
yep working on that

retrogtx commented:
> I have some UI suggestions. I think it would look cleaner if we don't show the 'Expires on' and 'Maximum number of uses' inputs and only have the gray text there:
> 
> ![Screenshot 2025-03-19 at 9 51 17 AM](https://private-user-images.githubusercontent.com/30310907/424384927-5940f7e0-9b2f-4c11-82b3-771753c18106.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI0NzI3MjgsIm5iZiI6MTc0MjQ3MjQyOCwicGF0aCI6Ii8zMDMxMDkwNy80MjQzODQ5MjctNTk0MGY3ZTAtOWIyZi00YzExLTgyYjMtNzcxNzUzYzE4MTA2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzIwVDEyMDcwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU0OWI1NzA4ZmEzZWE1ZjZmNWUyMGRmZmQ4ZWFiZDI4NTMzM2RhNzgyNzgyYWNjMmNhNmFjMTkwMGQ1YWY0NmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.GbqNg9cG033EST2qhkMvp04PxaSAnbWDa9XgXSnkrNI)
> 
> Instead, open a Dialog after clicking the settings button and handle all settings in there. In the dialog you can either pick 'Usage-based or 'Time-based', by default it's a usage-based linked with Number of uses = 1
> 
> Let me know what you think about that

I thought about that dialog as it was mentioned here soo did not go with the inline approach initially

alishaz-polymath commented:
> <img alt="image" width="1608" src="https://private-user-images.githubusercontent.com/76801862/424978604-fdd4cb15-07ba-46f9-bdee-f2bd130e36dc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI0NzQzMTIsIm5iZiI6MTc0MjQ3NDAxMiwicGF0aCI6Ii83NjgwMTg2Mi80MjQ5Nzg2MDQtZmRkNGNiMTUtMDdiYS00NmY5LWJkZWUtZjJiZDEzMGUzNmRjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzIwVDEyMzMzMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE5YzMwMWEwMWRhY2Q1YjU1NDViYmM1MGViMmI2NjYzZDZhZGY4ZmVhZWQwOGY1YTY1OWQ3NjBiNGU4OGM0ZGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ODgXFMV8z_C2t03MS5fAL8xY00P_kOi8mIYjRRoTGzk"> alright, is this nice?

I like this, deferring to @CarinaWolli for the final decision.

retrogtx committed:
refactor: update the dialogue layout a lil bit

retrogtx committed:
Merge branch 'main' into priv-links

retrogtx committed:
Merge branch 'main' into priv-links

retrogtx committed:
fix type / import err

retrogtx committed:
Merge branch 'main' into priv-links

retrogtx commented:
most of these are prettier errors from untouched files, strange

retrogtx committed:
fix imports

retrogtx committed:
Merge branch 'priv-links' of https://github.com/retrogtx/cal.com into priv-links

retrogtx committed:
make prettier happy

retrogtx committed:
Merge branch 'main' into priv-links

retrogtx committed:
Merge branch 'main' into priv-links

retrogtx committed:
Merge branch 'main' into priv-links

retrogtx committed:
Merge branch 'main' into priv-links

retrogtx committed:
refactor: every private link has the delete button now

make the settings background hover size same to the delete button

retrogtx committed:
fix prettier err

retrogtx committed:
resolve conflicts

retrogtx committed:
Merge branch 'main' into priv-links

retrogtx commented:
gentle ping, @alishaz-polymath

retrogtx committed:
Merge branch 'main' into priv-links

retrogtx committed:
refactor: remove unused variable

alishaz-polymath committed:
Merge branch 'main' into priv-links

CarinaWolli commented:
I have some design suggestions. 

Before: 
![Screenshot 2025-04-08 at 12 09 53 PM](https://github.com/user-attachments/assets/7ca4a0af-e84e-4a70-8820-6253d098804e)

After: 
![Screenshot 2025-04-08 at 12 04 32 PM](https://github.com/user-attachments/assets/4b35ba1d-d1fb-43b1-aaab-62afe1c8fbd1)
+ also make use of our radio button component, it should look like that: 
![Screenshot 2025-04-08 at 12 10 56 PM](https://github.com/user-attachments/assets/66126e2f-9398-4b00-a85f-cb52ea219e07)

retrogtx commented:
will be done 👍👍

CarinaWolli commented on apps/web/components/booking/CancelBooking.tsx:
```diff
@@ -199,23 +199,25 @@ export default function CancelBooking(props: Props) {
                 data-testid="confirm_cancel"
                 disabled={
                   props.isHost &&
+                  !props.seatReferenceUid &&
```
why are we adding this?

retrogtx replied:
Was working on a PR for seat based booking cancellations, mixed it up with this one 🫠

will remove after double checking once again

CarinaWolli commented on apps/web/modules/bookings/views/bookings-single-view.getServerSideProps.tsx:
```diff
@@ -169,14 +169,8 @@ export async function getServerSideProps(context: GetServerSidePropsContext) {
       bookingInfo?.user?.id === userId ||
-      eventType.users.some(
-        (user) =>
-          user.id === userId && bookingInfo.attendees.some((attendee) => attendee.email === user.email)
```
why was the attendee check removed?

CarinaWolli commented on packages/features/bookings/lib/handleNewBooking.ts:
```diff
@@ -1984,11 +1984,46 @@ async function handler(
+        const now = new Date();
+
+        if (hashedLink.expiresAt && hashedLink.expiresAt < now) {
+          throw new HttpError({ statusCode: 410, message: "Link has expired" });
```
@alishaz-polymath @hariombalhara what should happen when the link expired when confirming the booking but was still available when they visited the link? Same with usage based

alishaz-polymath replied:
I think we should have the same functionality as we do for other similar errors in the booking confirmation (say boundary limit or user not available).

CarinaWolli replied:
so show a 'Link expired' blue box error?

alishaz-polymath replied:
Yeah I think so.

CarinaWolli replied:
@retrogtx Can you add that? 🙏

retrogtx replied:
done!

CarinaWolli commented on packages/features/bookings/lib/handleNewBooking.ts:
```diff
@@ -1984,11 +1984,46 @@ async function handler(
+        }
+
+        if (shouldDeleteLink) {
+          await prisma.hashedLink.delete({
```
If hashed link is expired we never reach that code because with throw the error above https://github.com/calcom/cal.com/pull/20118/files#diff-a7c21727d9a0d4d1aaff9d24fca9293bc28468861bc8022dfa03116908b6fae7R1998

retrogtx committed:
new dialogue ui

retrogtx committed:
Merge branch 'main' into priv-links

retrogtx committed:
refactor: remove unused code for this PR

retrogtx committed:
refactor: remove changes meant for other PR

retrogtx committed:
fix: improve error handling / validation for hashed link usage in booking process

CarinaWolli commented on packages/trpc/server/routers/viewer/eventTypes/getHashedLink.handler.ts:
```diff
@@ -0,0 +1,90 @@
+    const membership = await prisma.membership.findFirst({
+      where: {
+        teamId: eventTypeTeamId,
+        userId,
```
```suggestion
        userId,
        accepted: true
```

CarinaWolli commented on packages/features/eventtypes/lib/hooks/useHashedLinkData.ts:
```diff
@@ -0,0 +1,45 @@
+import type { PrivateLinkWithOptions } from "@calcom/features/eventtypes/lib/types";
+import { trpc } from "@calcom/trpc/react";
+
+export const useHashedLinkData = (link: string) => {
```
I don't think this is used anywhere

retrogtx replied:
true thanks for pointing it out 🙏

CarinaWolli commented on packages/trpc/server/routers/viewer/eventTypes/getHashedLink.handler.ts:
```diff
@@ -0,0 +1,90 @@
+  };
+};
+
+export const getHashedLinkHandler = async ({ ctx, input }: GetHashedLinkOptions) => {
```
if we remove `useHashedLinkData` we can also remove this handler

retrogtx replied:
it's used in `getHashedLink` in `packages/trpc/server/routers/viewer/eventTypes/_router.ts`

CarinaWolli commented on packages/trpc/server/routers/viewer/eventTypes/getHashedLinks.handler.ts:
```diff
@@ -0,0 +1,78 @@
+        const membership = await prisma.membership.findFirst({
+          where: {
+            teamId: link.eventType.teamId,
+            userId,
```
```suggestion
            userId,
            accepted: true
```

CarinaWolli commented on packages/features/bookings/lib/handleNewBooking.ts:
```diff
@@ -1984,14 +1984,67 @@ async function handler(
+
+      if (hashedLink.maxUsageCount && hashedLink.maxUsageCount > 0) {
+        if (hashedLink.usageCount >= hashedLink.maxUsageCount) {
+          throw new HttpError({ statusCode: 410, message: "Link usage limit reached" });
```
```suggestion
        throw new HttpError({ statusCode: 410, message: "Link has expired" });
```

retrogtx committed:
Update packages/features/bookings/lib/handleNewBooking.ts

Co-authored-by: Carina Wollendorfer <30310907+CarinaWolli@users.noreply.github.com>

retrogtx committed:
Update packages/features/bookings/lib/handleNewBooking.ts

Co-authored-by: Carina Wollendorfer <30310907+CarinaWolli@users.noreply.github.com>

retrogtx committed:
fix: buggy rendering of remaining uses when adding a new private link

retrogtx committed:
refactor: remove useHashedLinkData hook and update handlers

retrogtx commented on packages/features/eventtypes/components/MultiplePrivateLinksController.tsx:
```diff
@@ -1,97 +1,313 @@
+                  } else {
+                    // Show remaining uses - if usedCount equals maxUses, we still have that last use
+                    linkDescription = `${remainingUses} of ${maxUses} ${
+                      remainingUses === 1 ? "use" : "uses"
```
fixed!

retrogtx commented on packages/features/eventtypes/components/HashedLinkUsageIndicator.tsx:
```diff
@@ -0,0 +1,41 @@
+
+import { trpc } from "@calcom/trpc/react";
+
+export const HashedLinkUsageIndicator = ({ link }: { link: string }) => {
```
done!

retrogtx committed:
refactor: remove HashedLinkUsageIndicator component and its usage

alishaz-polymath commented:
@retrogtx How is this looking? Is it ready for a review again?

alishaz-polymath commented:
> @alishaz-polymath it has always been ready!!

I mean have the mrge concerns been addressed?
If so, could you please resolve all comments that are resolved? As right now when I see them not resolved, I assume that there's work going on still. 🙏

retrogtx commented:
@alishaz-polymath yep they have been addressed, and i think that mrge bots comments are mostly fluff so resolving them - however if you feel like somethings off please let me know

i'd say this is good to go, been addressing comments over the past few weeks

alishaz-polymath committed:
Merge branch 'main' into priv-links

alishaz-polymath commented:
@retrogtx plenty of failing tests, can you take a look please?

retrogtx commented:
thats weird since they were all passing before updating

will look!

retrogtx committed:
fix type err

retrogtx commented:
fixed, @alishaz-polymath pinging you for the same 🙏

alishaz-polymath commented:
Awesome! Taking a look 🙏

alishaz-polymath reviewed (changes requested):
@retrogtx I'm seeing a behavior where the time based link is expiring on single use which seems strange. The time based link should only expire once the time is through. Can you please check?

retrogtx committed:
remove hashed link deletion logic in handleNewBooking.ts

retrogtx committed:
Merge branch 'main' into priv-links

retrogtx committed:
refactor: update e2e test for hash-my-url to clarify usage-based link behavior

retrogtx committed:
Merge branch 'main' into priv-links

retrogtx commented on packages/features/bookings/__tests__/private-link-booking.test.ts:
```diff
@@ -0,0 +1,176 @@
+import type { PrismaClient } from "@prisma/client";
```
thats not true, I can see it being used here:

`const prismaMock = mockDeep<PrismaClient>();`

retrogtx commented on packages/features/bookings/__tests__/private-link-booking.test.ts:
```diff
@@ -0,0 +1,176 @@
+  handleNewBooking: vi.fn(),
+}));
+
+const prismaMock = mockDeep<PrismaClient>();
```
it is being used here:

` mockReset(prismaMock);`

retrogtx commented on packages/features/bookings/__tests__/private-link-booking.test.ts:
```diff
@@ -0,0 +1,176 @@
+
+  describe("Time-based expiration", () => {
+    it("should throw error when booking with expired link", async () => {
+      const expiredLink = {
```
ah true

retrogtx committed:
remove unused variables, address mrge bot comments

retrogtx committed:
resolve type err

retrogtx commented:
@alishaz-polymath gentle ping!

alishaz-polymath commented:
@retrogtx Can you please close down the mrge comments which are resolved/unnecessary?

alishaz-polymath commented:
Now it is not expiring the single-use link. I could use it to book twice, and it's still not expired.

retrogtx committed:
fix null single link value

retrogtx committed:
Merge branch 'main' into priv-links

retrogtx commented:
@alishaz-polymath bossman 🫡

alishaz-polymath committed:
Merge remote-tracking branch 'origin/main' into priv-links

alishaz-polymath committed:
fix migration

alishaz-polymath committed:
combine migrations

alishaz-polymath committed:
more fixes

alishaz-polymath committed:
fix test

alishaz-polymath committed:
fix unnecessary changes

alishaz-polymath committed:
Merge branch 'main' into priv-links

alishaz-polymath commented:
Closing in favor of #22304
